### PR TITLE
Introduce `DefaultStatusCodeResponse`

### PR DESCRIPTION
**Pull request by TharmiganK**
Fixes: https://github.com/ballerina-platform/ballerina-library/issues/6491

## Examples

- Client side:

  ```bal
  http:StatusCodeClient clientEP = check new("localhost:9090/api");
  OkAlbumArray|http:DefaultStatusCodeResponse|error res = clientEP->/albums;
  
  type DefaultErrorResponse record {|
      *http:DefaultStatusCodeResponse;
      DefaultErrorMessage body;
      DefaultHeaders headers;
  |};
  OkAlbumArray|DefaultErrorResponse|error res = clientEP->/albums;
  ```

- Service side:

  ```bal
  resource function get default/[int statusCode]() returns http:DefaultStatusCodeResponse {
      if statusCode == 204 {
          return {status: new (204)};
      }
      return {
          body: "Default Response",
          status: new (statusCode)
      };
  }
  ```

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [ ] ~Updated the spec~ - will be addressed via https://github.com/ballerina-platform/ballerina-library/issues/6289
- [x] Checked native-image compatibility
- [x] Checked the impact on OpenAPI generation - will be addressed via https://github.com/ballerina-platform/ballerina-library/issues/6492

### CODE REVIEW

File: ballerina/http_status_code_types.bal [hunk 1]
```diff
@@ -602,6 +614,15 @@ final StatusNotExtended STATUS_NOT_EXTENDED_OBJ = new;
 final StatusNetworkAuthenticationRequired STATUS_NETWORK_AUTHENTICATION_REQUIRED_OBJ = new;
 
 // Status code record types
+
+# The default status code response record.
+#
+# + status - The response status code obj
```
Comment (dilanSachi) [inline ballerina/http_status_code_types.bal:620]:
```suggestion
# + status - The response status code object
```
Isn't it better to have the complete name?

**Review by dilanSachi — APPROVED** [review]
Added a minor comment. LGTM

**Review by dilanSachi — APPROVED** [review]

File: ballerina/http_client_endpoint.bal [hunk 2]
```diff
@@ -677,17 +677,41 @@ isolated function createResponseError(int statusCode, string reasonPhrase, map<s
...
     } else if 400 <= statusCode && statusCode <= 499 {
-        return error StatusCodeBindingClientRequestError(reasonPhrase, statusCode = statusCode, headers = headers, body = body);
+        return error StatusCodeBindingClientRequestError(reasonPhrase, statusCode = statusCode, headers = headers, body = body, fromDefaultStatusCodeMapping = false);
     } else {
-        return error StatusCodeBindingRemoteServerError(reasonPhrase, statusCode = statusCode, headers = headers, body = body);
+        return error StatusCodeBindingRemoteServerError(reasonPhrase, statusCode = statusCode, headers = headers, body = body, fromDefaultStatusCodeMapping = false);
+    }
+}
+
+isolated function createStatusCodeResponseDataBindingError("header"|"mediaType"|"payload"|"generic" errorType, boolean fromDefaultStatusCodeMapping,
```
Comment (ayeshLK) [inline ballerina/http_client_endpoint.bal:691]: Can't we create an enum or a type-def for error type and use it here ?

```ballerina
type DataBindingErrorType "header"|"mediaType"|"payload"|"generic";
```

or

```ballerina

enum DataBindingErrorType {
    HEADER, MEDIA_TYPE, PAYLOAD, GENERIC
}
```

File: ballerina/http_response.bal [hunk 3]
```diff
@@ -541,12 +541,12 @@ public class Response {
     }
 
     isolated function buildStatusCodeResponse(typedesc<anydata>? payloadType, typedesc<StatusCodeResponse> statusCodeResType,
-        boolean requireValidation, Status status, map<string|int|boolean|string[]|int[]|boolean[]> headers, string? mediaType)
-            returns StatusCodeResponse|ClientError {
+        boolean requireValidation, Status status, map<string|int|boolean|string[]|int[]|boolean[]> headers, string? mediaType,
+        boolean fromDefaultStatusCodeMapping) returns StatusCodeResponse|ClientError {
```
Comment (ayeshLK) [inline ballerina/http_response.bal:545]: This method is taking around 7 parameters, which reduce the readability significantly. Couldn't we refactor this method to something like this ?

```ballerina
type StatsCodeRespContext record {|
    typedesc<anydata> payloadType?;
    typedesc<StatusCodeResponse> statusCodeResType;
    boolean requireValidation; 
    Status status;
    map<string|int|boolean|string[]|int[]|boolean[]> headers;
    string mediaType?;
    boolean fromDefaultStatusCodeMapping;
|};

isolated function buildStatusCodeResponse(StatsCodeRespContext context) returns StatusCodeResponse|ClientError {
    // function logic here         
}
```

File: ballerina/http_response.bal [hunk 4]
```diff
@@ -562,18 +562,34 @@ public class Response {
...
+                return createStatusCodeResponseBindingError(statusCode, reasonPhrase, headers);
+            }
+            return error PayloadBindingClientError("http:StatusCodeBindingError creation failed: " + statusCode.toString() +
+                " response payload extraction failed", payload);
+        } else {
+            return createStatusCodeResponseBindingError(statusCode, reasonPhrase, headers, payload);
+        }
+    }
+
+    isolated function getStatusCodeResponseDataBindingError(string reasonPhrase, boolean fromDefaultStatusCodeMapping,
```
Comment (ayeshLK) [inline ballerina/http_response.bal:580]: Ditto. Refer this comment [1]

[1] - https://github.com/ballerina-platform/module-ballerina-http/pull/2002/files#r1616543125

File: native/src/main/java/io/ballerina/stdlib/http/api/nativeimpl/StatusCodeBindingException.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Status code binding exception to indicate that the status code data-binding has failed.
+ *
+ * @since 2.11.1
+ */
+public class StatusCodeBindingException extends RuntimeException {
+
+    final String errorType;
+    final BError cause;
```
Comment (ayeshLK) [inline native/src/main/java/io/ballerina/stdlib/http/api/nativeimpl/StatusCodeBindingException.java:30]:
```suggestion
    private final String errorType;
    private final BError cause;
```

File: ballerina/http_client_endpoint.bal (hunk 2 above)
Comment (TharmiganK) [reply ballerina/http_client_endpoint.bal:691]: This is used for internal purpose(non-public and called from java native function) and used only at this function. Thats why I have just defined it as the union of singletons

File: ballerina/http_response.bal (hunk 3 above)
Comment (TharmiganK) [reply ballerina/http_response.bal:545]: This is called from the java native function and used for internal purpose. Having a record means we need to build it from java native side and call this function, which adds an additional overhead for the function call

File: native/src/main/java/io/ballerina/stdlib/http/api/nativeimpl/StatusCodeBindingException.java (hunk 5 above)
Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/http/api/nativeimpl/StatusCodeBindingException.java]: Fixed via [3ed8b32](https://github.com/ballerina-platform/module-ballerina-http/pull/2002/commits/3ed8b32953953b57dc0074a09681ce99e755e232)

File: ballerina/http_response.bal (hunk 4 above)
Comment (TharmiganK) [reply ballerina/http_response.bal:580]: Fixed via [3ed8b32](https://github.com/ballerina-platform/module-ballerina-http/pull/2002/commits/3ed8b32953953b57dc0074a09681ce99e755e232)

File: ballerina/http_client_endpoint.bal (hunk 2 above)
Comment (TharmiganK) [reply ballerina/http_client_endpoint.bal]: Fixed via [3ed8b32](https://github.com/ballerina-platform/module-ballerina-http/pull/2002/commits/3ed8b32953953b57dc0074a09681ce99e755e232)

File: ballerina/http_response.bal (hunk 3 above)
Comment (ayeshLK) [reply ballerina/http_response.bal:545]: What if we pass this as an included record param ?
```ballerina
isolated function buildStatusCodeResponse(*StatsCodeRespContext context) returns StatusCodeResponse|ClientError;
```

Would this fix it ?

Comment (TharmiganK) [reply ballerina/http_response.bal:545]: For included parameter, we have to pass it as a BMap from the Java side. So this is same as having a record parameter

**Review by lnash94 — APPROVED** [review]

**Pull request merged**